File: plugin/src/main/java/com/stratio/cassandra/lucene/IndexOptions.java
/*
 * Licensed to STRATIO (C) under one or more contributor license agreements.
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.  The STRATIO (C) licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.stratio.cassandra.lucene;

import com.google.common.base.MoreObjects;
import com.stratio.cassandra.lucene.schema.Schema;
import com.stratio.cassandra.lucene.schema.SchemaBuilder;
import org.apache.cassandra.config.CFMetaData;
import org.apache.cassandra.db.Directories;
import org.apache.cassandra.schema.IndexMetadata;

import java.io.File;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * Lucene {@link Index} user-specified configuration options parser.
 *
 * @author Andres de la Pena {@literal <adelapena@stratio.com>}
 */
public class IndexOptions {

    public static final String REFRESH_SECONDS_OPTION = "refresh_seconds";
    public static final double DEFAULT_REFRESH_SECONDS = 60;

    public static final String RAM_BUFFER_MB_OPTION = "ram_buffer_mb";
    public static final int DEFAULT_RAM_BUFFER_MB = 64;

    public static final String MAX_MERGE_MB_OPTION = "max_merge_mb";
    public static final int DEFAULT_MAX_MERGE_MB = 5;

    public static final String MAX_CACHED_MB_OPTION = "max_cached_mb";
    public static final int DEFAULT_MAX_CACHED_MB = 30;

    public static final String INDEXING_THREADS_OPTION = "indexing_threads";
    public static final int DEFAULT_INDEXING_THREADS = 0;

    public static final String INDEXING_QUEUES_SIZE_OPTION = "indexing_queues_size";
    public static final int DEFAULT_INDEXING_QUEUES_SIZE = 50;

    public static final String EXCLUDED_DATA_CENTERS_OPTION = "excluded_data_centers";
    public static final List<String> DEFAULT_EXCLUDED_DATA_CENTERS = Collections.emptyList();

    public static final String TOKEN_RANGE_CACHE_SIZE_OPTION = "token_range_cache_size";
    public static final int DEFAULT_TOKEN_RANGE_CACHE_SIZE = 16;

    public static final String SEARCH_CACHE_SIZE_OPTION = "search_cache_size";
    public static final int DEFAULT_SEARCH_CACHE_SIZE = 16;

    public static final String DIRECTORY_PATH_OPTION = "directory_path";
    public static final String INDEXES_DIR_NAME = "lucene";

    public static final String SCHEMA_OPTION = "schema";

    /** The mapping schema */
    public final Schema schema;

    /** The path of the directory where the index files will be stored */
    public final Path path;

    /** The Lucene index searcher refresh frequency, in seconds */
    public final double refreshSeconds;

    /** The Lucene's max RAM buffer size, in MB */
    public final int ramBufferMB;

    /** The Lucene's max segments merge size size, in MB */
    public final int maxMergeMB;

    /** The Lucene's max cache size, in MB */
    public final int maxCachedMB;

    /** The number of asynchronous indexing threads */
    public final int indexingThreads;

    /** The size of the asynchronous indexing queues */
    public final int indexingQueuesSize;

    /** The max size of the token range cache */
    public final int tokenRangeCacheSize;

    /** The max size of the search cache */
    public final int searchCacheSize;

    /** The names of the data centers excluded from indexing */
    public final List<String> excludedDataCenters;

    /**
     * Builds a new {@link IndexOptions} for the column family and index metadata.
     *
     * @param tableMetadata the indexed table metadata
     * @param indexMetadata the index metadata
     */
    public IndexOptions(CFMetaData tableMetadata, IndexMetadata indexMetadata) {
        Map<String, String> options = indexMetadata.options;
        refreshSeconds = parseRefresh(options);
        ramBufferMB = parseRamBufferMB(options);
        maxMergeMB = parseMaxMergeMB(options);
        maxCachedMB = parseMaxCachedMB(options);
        indexingThreads = parseIndexingThreads(options);
        indexingQueuesSize = parseIndexingQueuesSize(options);
        excludedDataCenters = parseExcludedDataCenters(options);
        tokenRangeCacheSize = parseTokenRangeCacheSize(options);
        searchCacheSize = parseSearchCacheSize(options);
        path = parsePath(options, tableMetadata, indexMetadata);
        schema = parseSchema(options, tableMetadata);
    }

    /**
     * Validates the specified index options.
     *
     * @param options the options to be validated
     * @param metadata the indexed table metadata
     */
    public static void validateOptions(Map<String, String> options, CFMetaData metadata) {
        parseRefresh(options);
        parseRamBufferMB(options);
        parseMaxMergeMB(options);
        parseMaxCachedMB(options);
        parseIndexingThreads(options);
        parseIndexingQueuesSize(options);
        parseExcludedDataCenters(options);
        parseTokenRangeCacheSize(options);
        parseSearchCacheSize(options);
        parseSchema(options, metadata);
        parsePath(options, metadata, null);
    }

    private static double parseRefresh(Map<String, String> options) {
        String refreshOption = options.get(REFRESH_SECONDS_OPTION);
        if (refreshOption != null) {
            double refreshSeconds;
            try {
                refreshSeconds = Double.parseDouble(refreshOption);
            } catch (NumberFormatException e) {
                throw new IndexException("'%s' must be a strictly positive double", REFRESH_SECONDS_OPTION);
            }
            if (refreshSeconds <= 0) {
                throw new IndexException("'%s' must be strictly positive", REFRESH_SECONDS_OPTION);
            }
            return refreshSeconds;
        } else {
            return DEFAULT_REFRESH_SECONDS;
        }
    }

    private static int parseRamBufferMB(Map<String, String> options) {
        String ramBufferSizeOption = options.get(RAM_BUFFER_MB_OPTION);
        if (ramBufferSizeOption != null) {
            int ramBufferMB;
            try {
                ramBufferMB = Integer.parseInt(ramBufferSizeOption);
            } catch (NumberFormatException e) {
                throw new IndexException("'%s' must be a strictly positive integer", RAM_BUFFER_MB_OPTION);
            }
            if (ramBufferMB <= 0) {
                throw new IndexException("'%s' must be strictly positive", RAM_BUFFER_MB_OPTION);
            }
            return ramBufferMB;
        } else {
            return DEFAULT_RAM_BUFFER_MB;
        }
    }

    private static int parseMaxMergeMB(Map<String, String> options) {
        String maxMergeSizeMBOption = options.get(MAX_MERGE_MB_OPTION);
        if (maxMergeSizeMBOption != null) {
            int maxMergeMB;
            try {
                maxMergeMB = Integer.parseInt(maxMergeSizeMBOption);
            } catch (NumberFormatException e) {
                throw new IndexException("'%s' must be a strictly positive integer", MAX_MERGE_MB_OPTION);
            }
            if (maxMergeMB <= 0) {
                throw new IndexException("'%s' must be strictly positive", MAX_MERGE_MB_OPTION);
            }
            return maxMergeMB;
        } else {
            return DEFAULT_MAX_MERGE_MB;
        }
    }

    private static int parseMaxCachedMB(Map<String, String> options) {
        String maxCachedMBOption = options.get(MAX_CACHED_MB_OPTION);
        if (maxCachedMBOption != null) {
            int maxCachedMB;
            try {
                maxCachedMB = Integer.parseInt(maxCachedMBOption);
            } catch (NumberFormatException e) {
                throw new IndexException("'%s' must be a strictly positive integer", MAX_CACHED_MB_OPTION);
            }
            if (maxCachedMB <= 0) {
                throw new IndexException("'%s' must be strictly positive", MAX_CACHED_MB_OPTION);
            }
            return maxCachedMB;
        } else {
            return DEFAULT_MAX_CACHED_MB;
        }
    }

    private static int parseIndexingThreads(Map<String, String> options) {
        String indexPoolNumQueuesOption = options.get(INDEXING_THREADS_OPTION);
        if (indexPoolNumQueuesOption != null) {
            try {
                return Integer.parseInt(indexPoolNumQueuesOption);
            } catch (NumberFormatException e) {
                throw new IndexException("'%s' must be a positive integer", INDEXING_THREADS_OPTION);
            }
        } else {
            return DEFAULT_INDEXING_THREADS;
        }
    }

    private static int parseIndexingQueuesSize(Map<String, String> options) {
        String indexPoolQueuesSizeOption = options.get(INDEXING_QUEUES_SIZE_OPTION);
        if (indexPoolQueuesSizeOption != null) {
            int indexingQueuesSize;
            try {
                indexingQueuesSize = Integer.parseInt(indexPoolQueuesSizeOption);
            } catch (NumberFormatException e) {
                throw new IndexException("'%s' must be a strictly positive integer", INDEXING_QUEUES_SIZE_OPTION);
            }
            if (indexingQueuesSize <= 0) {
                throw new IndexException("'%s' must be strictly positive", INDEXING_QUEUES_SIZE_OPTION);
            }
            return indexingQueuesSize;
        } else {
            return DEFAULT_INDEXING_QUEUES_SIZE;
        }
    }

    private static List<String> parseExcludedDataCenters(Map<String, String> options) {
        String excludedDataCentersOption = options.get(EXCLUDED_DATA_CENTERS_OPTION);
        if (excludedDataCentersOption != null) {
            String[] array = excludedDataCentersOption.trim().split(",");
            return Arrays.asList(array);
        } else {
            return DEFAULT_EXCLUDED_DATA_CENTERS;
        }
    }

    private static int parseTokenRangeCacheSize(Map<String, String> options) {
        String tokenRangeCacheSizeOption = options.get(TOKEN_RANGE_CACHE_SIZE_OPTION);
        if (tokenRangeCacheSizeOption != null) {
            int tokenRangeCacheSize;
            try {
                tokenRangeCacheSize = Integer.parseInt(tokenRangeCacheSizeOption);
            } catch (NumberFormatException e) {
                throw new IndexException("'%s' must be a positive integer", TOKEN_RANGE_CACHE_SIZE_OPTION);
            }
            if (tokenRangeCacheSize < 0) {
                throw new IndexException("'%s' must be positive", TOKEN_RANGE_CACHE_SIZE_OPTION);
            }
            return tokenRangeCacheSize;
        } else {
            return DEFAULT_TOKEN_RANGE_CACHE_SIZE;
        }
    }

    private static int parseSearchCacheSize(Map<String, String> options) {
        String searchCacheSizeOption = options.get(SEARCH_CACHE_SIZE_OPTION);
        if (searchCacheSizeOption != null) {
            int searchCacheSize;
            try {
                searchCacheSize = Integer.parseInt(searchCacheSizeOption);
            } catch (NumberFormatException e) {
                throw new IndexException("'%s' must be a positive integer", SEARCH_CACHE_SIZE_OPTION);
            }
            if (searchCacheSize < 0) {
                throw new IndexException("'%s' must be positive", SEARCH_CACHE_SIZE_OPTION);
            }
            return searchCacheSize;
        } else {
            return DEFAULT_SEARCH_CACHE_SIZE;
        }
    }

    private static Path parsePath(Map<String, String> options, CFMetaData tableMetadata, IndexMetadata indexMetadata) {
        String pathOption = options.get(DIRECTORY_PATH_OPTION);
        if (pathOption != null) {
            return Paths.get(pathOption);
        } else if (indexMetadata != null) {
            Directories directories = new Directories(tableMetadata);
            String basePath = directories.getDirectoryForNewSSTables().getAbsolutePath();
            return Paths.get(basePath + File.separator + INDEXES_DIR_NAME + File.separator + indexMetadata.name);
        }
        return null;
    }

    private static Schema parseSchema(Map<String, String> options, CFMetaData tableMetadata) {
        String schemaOption = options.get(SCHEMA_OPTION);
        if (schemaOption != null && !schemaOption.trim().isEmpty()) {
            Schema schema;
            try {
                schema = SchemaBuilder.fromJson(schemaOption).build();
                schema.validate(tableMetadata);
                return schema;
            } catch (Exception e) {
                throw new IndexException(e, "'%s' is invalid : %s", SCHEMA_OPTION, e.getMessage());
            }
        } else {
            throw new IndexException("'%s' required", SCHEMA_OPTION);
        }
    }

    /** {@inheritDoc} */
    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                          .add("refreshSeconds", refreshSeconds)
                          .add("ramBufferMB", ramBufferMB)
                          .add("maxMergeMB", maxMergeMB)
                          .add("maxCachedMB", maxCachedMB)
                          .add("indexingThreads", indexingThreads)
                          .add("indexingQueuesSize", indexingQueuesSize)
                          .add("excludedDataCenters", excludedDataCenters)
                          .add("tokenRangeCacheSize", tokenRangeCacheSize)
                          .add("searchCacheSize", searchCacheSize)
                          .add("path", path)
                          .add("schema", schema)
                          .toString();
    }
}

File: plugin/src/main/scala/com/stratio/cassandra/lucene/IndexOptions.scala
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene

import java.nio.file.{Path, Paths}

import com.stratio.cassandra.lucene.IndexOptions._
import com.stratio.cassandra.lucene.partitioning.{Partitioner, PartitionerOnNone}
import com.stratio.cassandra.lucene.schema.{Schema, SchemaBuilder}
import com.stratio.cassandra.lucene.util.SchemaValidator
import org.apache.cassandra.config.{CFMetaData, DatabaseDescriptor}
import org.apache.cassandra.db.Directories
import org.apache.cassandra.schema.IndexMetadata

import scala.collection.JavaConverters._

/** Index user-specified configuration options parser.
  *
  * @param tableMetadata the indexed table metadata
  * @param indexMetadata the index metadata
  * @author Andres de la Pena `adelapena@stratio.com`
  */
class IndexOptions(tableMetadata: CFMetaData, indexMetadata: IndexMetadata) {

  val options = indexMetadata.options.asScala.toMap

  /** The Lucene index searcher refresh frequency, in seconds */
  val refreshSeconds = parseRefresh(options)

  /** The Lucene's max RAM buffer size, in MB */
  val ramBufferMB = parseRamBufferMB(options)

  /** The Lucene's max segments merge size size, in MB */
  val maxMergeMB = parseMaxMergeMB(options)

  /** The Lucene's max cache size, in MB */
  val maxCachedMB = parseMaxCachedMB(options)

  /** The number of asynchronous indexing threads */
  val indexingThreads = parseIndexingThreads(options)

  /** The size of the asynchronous indexing queues */
  val indexingQueuesSize = parseIndexingQueuesSize(options)

  /** The names of the data centers excluded from indexing */
  val excludedDataCenters = parseExcludedDataCenters(options)

  /** The mapping schema */
  val schema = parseSchema(options, tableMetadata)

  val pathAndPartitioner = parsePathAndPartitioner(options,
    tableMetadata,
    DatabaseDescriptor.getAllDataFileLocations.map(Paths.get(_)),
    getBaseTablePath(tableMetadata))

  /** The path of the directory where the index files will be stored */
  val path = pathAndPartitioner._1

  /** The index partitioner */
  val partitioner = pathAndPartitioner._2

  /** If the index is sparse or not */
  val sparse = parseSparse(options, tableMetadata)
}

/** Companion object for [[IndexOptions]]. */
object IndexOptions {

  val REFRESH_SECONDS_OPTION = "refresh_seconds"
  val DEFAULT_REFRESH_SECONDS = 60D

  val RAM_BUFFER_MB_OPTION = "ram_buffer_mb"
  val DEFAULT_RAM_BUFFER_MB = 64

  val MAX_MERGE_MB_OPTION = "max_merge_mb"
  val DEFAULT_MAX_MERGE_MB = 5

  val MAX_CACHED_MB_OPTION = "max_cached_mb"
  val DEFAULT_MAX_CACHED_MB = 30

  val INDEXING_THREADS_OPTION = "indexing_threads"
  val DEFAULT_INDEXING_THREADS = Runtime.getRuntime.availableProcessors

  val INDEXING_QUEUES_SIZE_OPTION = "indexing_queues_size"
  val DEFAULT_INDEXING_QUEUES_SIZE = 50

  val EXCLUDED_DATA_CENTERS_OPTION = "excluded_data_centers"
  val DEFAULT_EXCLUDED_DATA_CENTERS = List[String]()

  val DIRECTORY_PATH_OPTION = "directory_path"
  val INDEXES_DIR_NAME = "lucene"

  val SCHEMA_OPTION = "schema"

  val PARTITIONER_OPTION = "partitioner"
  val DEFAULT_PARTITIONER = PartitionerOnNone()

  val SPARSE_OPTION = "sparse"
  val DEFAULT_SPARSE = false

  /** Validates the specified index options.
    *
    * @param options  the options to be validated
    * @param metadata the indexed table metadata
    */
  def validate(options: java.util.Map[String, String], metadata: CFMetaData) {
    val o = options.asScala.toMap
    parseRefresh(o)
    parseRamBufferMB(o)
    parseMaxMergeMB(o)
    parseMaxCachedMB(o)
    parseIndexingThreads(o)
    parseIndexingQueuesSize(o)
    parseExcludedDataCenters(o)
    parseSchema(o, metadata)
    parsePathAndPartitioner(o,
      metadata,
      DatabaseDescriptor.getAllDataFileLocations.map(Paths.get(_)),
      getBaseTablePath(metadata))
  }

  def parseRefresh(options: Map[String, String]): Double = {
    parseStrictlyPositiveDouble(options, REFRESH_SECONDS_OPTION, DEFAULT_REFRESH_SECONDS)
  }

  private def parseStrictlyPositiveDouble(
      options: Map[String, String],
      name: String,
      default: Double): Double = {
    options.get(name).map(
      string => try string.toDouble catch {
        case e: NumberFormatException =>
          throw new IndexException(s"'$name' must be a strictly positive decimal, found: $string")
      }).map(
      double => if (double > 0) double
      else {
        throw new IndexException(s"'$name' must be strictly positive, found: $double")
      }).getOrElse(default)
  }

  def parseRamBufferMB(options: Map[String, String]): Int = {
    parseStrictlyPositiveInt(options, RAM_BUFFER_MB_OPTION, DEFAULT_RAM_BUFFER_MB)
  }

  def parseMaxMergeMB(options: Map[String, String]): Int = {
    parseStrictlyPositiveInt(options, MAX_MERGE_MB_OPTION, DEFAULT_MAX_MERGE_MB)
  }

  def parseMaxCachedMB(options: Map[String, String]): Int = {
    parseStrictlyPositiveInt(options, MAX_CACHED_MB_OPTION, DEFAULT_MAX_CACHED_MB)
  }

  private def parseStrictlyPositiveInt(
      options: Map[String, String],
      name: String,
      default: Int): Int = {
    options.get(name).map(
      string => try string.toInt catch {
        case e: NumberFormatException =>
          throw new IndexException(s"'$name' must be a strictly positive integer, found: $string")
      }).map(
      integer => if (integer > 0) integer
      else {
        throw new IndexException(s"'$name' must be strictly positive, found: $integer")
      }).getOrElse(default)
  }

  def parseIndexingThreads(options: Map[String, String]): Int = {
    parseInt(options, INDEXING_THREADS_OPTION, DEFAULT_INDEXING_THREADS)
  }

  private def parseInt(options: Map[String, String], name: String, default: Int): Int = {
    options.get(name).map(
      string => try string.toInt catch {
        case e: NumberFormatException =>
          throw new IndexException(s"'$name' must be an integer, found: $string")
      }).getOrElse(default)
  }

  def parseIndexingQueuesSize(options: Map[String, String]): Int = {
    parseStrictlyPositiveInt(options, INDEXING_QUEUES_SIZE_OPTION, DEFAULT_INDEXING_QUEUES_SIZE)
  }

  def parseExcludedDataCenters(options: Map[String, String]): List[String] = {
    options
      .get(EXCLUDED_DATA_CENTERS_OPTION)
      .map(_.split(",").map(_.trim).filterNot(_.isEmpty).toList)
      .getOrElse(DEFAULT_EXCLUDED_DATA_CENTERS)
  }

  def parsePathAndPartitioner(
      options: Map[String, String],
      table: CFMetaData,
      cassandraPathDirs: Array[Path],
      baseTablePath: Path): (Option[Path], Partitioner) = {
    var path = parsePath(options)
    val partitioner = parsePartitioner(options, table)
    val customPartitionerPaths = partitioner.pathsForEachPartitions
    if (cassandraPathDirs.length > 1) {
      if (customPartitionerPaths.isDefined) {
        for (cassandraFile <- cassandraPathDirs) {
          for (partitionerPath <- customPartitionerPaths.get) {
            if (partitionerPath.startsWith(cassandraFile)) {
              throw new IndexException(s"When cassandra is configured with more than one 'data_file_directory', custom partitioner paths must not be inside any of those 'data_file_directory','$partitionerPath' is inside: '$cassandraFile'")
            }
          }
        }
      } else {
        if (path.isDefined) {
          for (cassandraFile <- cassandraPathDirs) {
            if (path.get.startsWith(cassandraFile)) {
              throw new IndexException(s"When cassandra is configured with more than one 'data_file_directory', 'directory_path' must not be inside any of those 'data_file_directory','${path.get}' is inside: '$cassandraFile'")
            }
          }
        } else {
          throw new IndexException(s"When cassandra is configured with more than one 'data_file_directory', 'directory_path' required")
        }
      }
    } else if ((cassandraPathDirs.length == 1) && customPartitionerPaths.isEmpty && path.isEmpty) {
      path = Some(baseTablePath)
    }
    (path, partitioner)
  }

  def parsePartitioner(options: Map[String, String], table: CFMetaData): Partitioner = {
    options.get(PARTITIONER_OPTION).map(
      value => try {
        Partitioner.fromJson(table, value)
      } catch {
        case e: Exception => throw new IndexException(e,
          s"'$PARTITIONER_OPTION' is invalid : ${e.getMessage}")
      }).getOrElse(DEFAULT_PARTITIONER)
  }

  def parsePath(options: Map[String, String]): Option[Path] =
    options.get(DIRECTORY_PATH_OPTION).map(Paths.get(_)).orElse(None)

  def parseSchema(options: Map[String, String], table: CFMetaData): Schema = {
    options.get(SCHEMA_OPTION).map(
      value => try {
        val schema = SchemaBuilder.fromJson(value).build
        SchemaValidator.validate(schema, table)
        schema
      } catch {
        case e: Exception => throw new IndexException(e,
          s"'$SCHEMA_OPTION' is invalid : ${e.getMessage}")
      }).getOrElse(throw new IndexException(s"'$SCHEMA_OPTION' is required"))
  }

  def getBaseTablePath(table: CFMetaData): Path =
    Paths.get(new Directories(table).getDirectoryForNewSSTables.getAbsolutePath)

  def parseSparse(options: Map[String, String], table: CFMetaData): Boolean = {
    options.get(SPARSE_OPTION).map(
      value => try value.toBoolean catch {
        case e: Exception => throw new IndexException(e,
          s"'$SPARSE_OPTION' is invalid : ${e.getMessage}")
      }).getOrElse(DEFAULT_SPARSE)
  }
}
